File: network/src/main/java/moonbox/network/server/TransportServer.java
package moonbox.network.server;

import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelInitializer;
import io.netty.channel.ChannelOption;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.util.concurrent.DefaultThreadFactory;
import moonbox.network.TransportContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Closeable;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.concurrent.TimeUnit;

public class TransportServer implements Closeable {
  private static final Logger logger = LoggerFactory.getLogger(TransportServer.class);

  private TransportContext context;
  private ServerBootstrap bootstrap;
  private ChannelFuture channelFuture;
  private RpcHandler appRpcHandler;
  private String hostToBind;
  private int portToBind = -1;


  public TransportServer(
      TransportContext context,
      String hostToBind,
      int portToBind,
      RpcHandler appRpcHandler) {
    this.context = context;
    this.hostToBind = hostToBind;
    this.portToBind = portToBind;
    this.appRpcHandler = appRpcHandler;
  }

  public int start() throws IOException {
    try {
      this.portToBind = init(hostToBind, portToBind);
    } catch (RuntimeException e) {
      try {
        close();
      } catch (IOException ex) {
        logger.error("IOException should not have been thrown.", ex);
        throw ex;
      }
    }
    return this.portToBind;
  }

  private int init(String hostToBind, int portToBind) {
    EventLoopGroup boosGroup = new NioEventLoopGroup(0, new DefaultThreadFactory(this.getClass(), true));
    EventLoopGroup workerGroup = boosGroup;
    bootstrap = new ServerBootstrap()
        .group(boosGroup, workerGroup)
        .channel(NioServerSocketChannel.class)
        .option(ChannelOption.SO_BACKLOG, 1024)
        .childOption(ChannelOption.SO_KEEPALIVE, true)
        .childOption(ChannelOption.SO_RCVBUF, 10240)
        .childOption(ChannelOption.SO_SNDBUF, 10240)
        .childHandler(new ChannelInitializer<SocketChannel>() {
          @Override
          protected void initChannel(SocketChannel ch) throws Exception {
            context.initializePipeline(ch, appRpcHandler);
          }
        });
    InetSocketAddress address = hostToBind == null ?
        new InetSocketAddress(portToBind): new InetSocketAddress(hostToBind, portToBind);
    channelFuture = bootstrap.bind(address);
    channelFuture.syncUninterruptibly();

    return  ((InetSocketAddress) channelFuture.channel().localAddress()).getPort();
  }

  public int getPort() {
    if (portToBind == -1) {
      throw new IllegalStateException("Server not initialized");
    }
    return portToBind;
  }

  @Override
  public void close() throws IOException {
    if (channelFuture != null) {
      channelFuture.channel().close().awaitUninterruptibly(10, TimeUnit.SECONDS);
      channelFuture = null;
    }
    if (bootstrap != null && bootstrap.group() != null) {
      bootstrap.group().shutdownGracefully();
    }
    if (bootstrap != null && bootstrap.childGroup() != null) {
      bootstrap.childGroup().shutdownGracefully();
    }
    bootstrap = null;
  }

}

File: network/src/main/scala/moonbox/network/server/TransportServer.scala
/*-
 * <<
 * Moonbox
 * ==
 * Copyright (C) 2016 - 2019 EDP
 * ==
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * >>
 */

package moonbox.network.server

import java.net.InetSocketAddress
import java.util.concurrent.TimeUnit

import com.google.protobuf.MessageLite
import io.netty.bootstrap.ServerBootstrap
import io.netty.channel._
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioServerSocketChannel
import io.netty.handler.codec.protobuf.{ProtobufDecoder, ProtobufEncoder, ProtobufVarint32FrameDecoder, ProtobufVarint32LengthFieldPrepender}
import io.netty.util.concurrent.DefaultThreadFactory
import moonbox.common.{MbConf, MbLogging}
import moonbox.network.config.PORT_MAX_RETRY_TIMES

class TransportServer(host: String, port: Int, conf: MbConf, messageLite: MessageLite, handlers: List[ChannelHandler]) extends MbLogging {

	private var channelFuture: ChannelFuture = _
	private var bootstrap: ServerBootstrap = _

	def start(): Int = {

		val eventLoopGroup = new NioEventLoopGroup(0, new DefaultThreadFactory(this.getClass, true))
		bootstrap = new ServerBootstrap()
				.group(eventLoopGroup, eventLoopGroup)
				.channel(classOf[NioServerSocketChannel])
				.option(ChannelOption.SO_BACKLOG, Integer.valueOf(1024))
				.childOption(ChannelOption.SO_KEEPALIVE, java.lang.Boolean.valueOf(true))
				.childOption(ChannelOption.SO_RCVBUF, Integer.valueOf(10240))
				.childOption(ChannelOption.SO_SNDBUF, Integer.valueOf(10240))
				.childHandler(new ChannelInitializer[SocketChannel] {
					override def initChannel(channel: SocketChannel): Unit = {
						channel.pipeline
								.addLast(new ProtobufVarint32FrameDecoder())
								.addLast("decoder", new ProtobufDecoder(messageLite))
								.addLast(new ProtobufVarint32LengthFieldPrepender())
								.addLast("encoder", new ProtobufEncoder())
								.addLast(handlers: _*)
					}
				})

		// Bind and start to accept incoming connections.
		logInfo("Starting Master Transport server ...")
		val maxRetries = conf.get(PORT_MAX_RETRY_TIMES)
		for (offset <- 0 to maxRetries) {
			val tryPort = if (port == 0) port else port + offset
			try {
				channelFuture = bootstrap.bind(host, tryPort)
				channelFuture.syncUninterruptibly()
				logInfo(s"Master Transport server is listening on $host:$tryPort.")
				return tryPort
			} catch {
				case e: Exception =>
					if (offset >= maxRetries) {
						throw e
					}
					if (port == 0) {
						logWarning(s"TransportServer could not bind on a random free port. Attempting again.")
					} else {
						logWarning(s"TransportServer could not bind on port $tryPort. Attempting port ${tryPort + 1}")
					}
			}
		}

		channelFuture.channel().localAddress().asInstanceOf[InetSocketAddress].getPort
	}

	def stop(): Unit = {
		if (channelFuture != null) {
			channelFuture.channel().close().awaitUninterruptibly(10, TimeUnit.SECONDS)
			channelFuture = null
		}
		if (bootstrap != null && bootstrap.group() != null) {
			bootstrap.group().shutdownGracefully()
		}
		if (bootstrap != null && bootstrap.childGroup() != null) {
			bootstrap.childGroup().shutdownGracefully()
		}
		bootstrap = null
	}

}
